presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteAggregationIfToFilter.java

```diff
@@ -0,0 +1,197 @@
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.IF;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.ImmutableSortedMap.toImmutableSortedMap;
+import static java.util.function.Function.identity;
+
+/**
+ * A optimizer rule which rewrites
+ * AGG(IF(condition, expr))
+ * to
+ * AGG(expr) FILTER (WHERE condition).
+ * <p>
+ * The latter plan is more efficient because:
+ * 1. The filter can be pushed down to the scan node.
+ * 2. The rows not matching the condition are not aggregated.
+ * 3. The IF() expression wrapper is removed.
+ */
+public class RewriteAggregationIfToFilter
+        implements Rule<AggregationNode>
+{
+    private static final Capture<ProjectNode> CHILD = newCapture();
+
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .with(source().matching(project().capturedAs(CHILD)));
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isAggregationIfToFilterRewriteEnabled(session);
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        ProjectNode sourceProject = captures.get(CHILD);
+
+        Set<Aggregation> aggregationsToRewrite = aggregationNode.getAggregations().values().stream()
+                .filter(aggregation -> shouldRewriteAggregation(aggregation, sourceProject))
+                .collect(toImmutableSet());
+        if (aggregationsToRewrite.isEmpty()) {
+            return Result.empty();
+        }
+
+        // Get the corresponding assignments in the input project.
+        // The aggregationReferences only has the aggregations to rewrite, thus the sourceAssignments only has IF expressions with NULL false results.
+        // Multiple aggregations may reference the same input. We use a map to dedup them based on the VariableReferenceExpression, so that we only do the rewrite once per input
+        // IF expression.
+        // The order of sourceAssignments determines the order of generating the new variables for the IF conditions and results. We use a sorted map to get a deterministic
+        // order based on the name of the VariableReferenceExpressions.
+        Map<VariableReferenceExpression, RowExpression> sourceAssignments = aggregationsToRewrite.stream()
+                .map(a -> (VariableReferenceExpression) a.getArguments().get(0))
+                .collect(toImmutableSortedMap(VariableReferenceExpression::compareTo, identity(), v -> sourceProject.getAssignments().get(v), (v1, v2) -> v1));
```

Comment:
nit: break the params of `toImmutableSortedMap` to 4 lines for readability:
- replace `v` with `variable`
- replace `v1` and `v2` with `left` and `right`